### PR TITLE
Split Cerberus into modules

### PR DESCRIPTION
**Pull request by iainrb**
- Split Cerberus into four modules: core, cli, client, webapp
- Build a JAR including dependencies for cli, to allow easier running
- Additional copies of some mock test data, so each module has correct test resources
- Reinstate /provenance URI suffix for ProvenanceServlet.java; avoids conflict with other servlets
- Improved argument checking for cli
- Update README

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,8 @@
 /target/
 /nbproject/
 Temp*
-*~
\ No newline at end of file
+*~
```
Comment (apmasell) [inline .gitignore:4]: `target` is preferable to absolute paths

File: cerberus-cli/src/main/java/ca/on/oicr/gsi/cerberus/Cerberus.java [hunk 2]
```diff
@@ -61,35 +71,79 @@ public static void main(String[] args) {
...
         }
 
-        if (line.hasOption("help")) {
-            new HelpFormatter().printHelp("java Cerberus", options);
+        if (line.hasOption(HELP_OPT)) {
+            new HelpFormatter().printHelp("java -jar cerberus-cli/target/cerberus-cli-${VERSION}-jar-with-dependencies.jar ", options);
             System.exit(0);
         }
 
+        Boolean error = false;
```
Comment (apmasell) [inline cerberus-cli/src/main/java/ca/on/oicr/gsi/cerberus/Cerberus.java]: `boolean` primitive preferred to object `Boolean` type.

File: cerberus-cli/src/main/java/ca/on/oicr/gsi/cerberus/Cerberus.java [hunk 3]
```diff
@@ -61,35 +71,79 @@ public static void main(String[] args) {
...
+        // check required options are present
+        ArrayList<String> required = new ArrayList<>();
```
Comment (apmasell) [inline cerberus-cli/src/main/java/ca/on/oicr/gsi/cerberus/Cerberus.java]: By Java convention, it's preferred to use the `List` and `Map` interfaces even when filling it with a particular subclass.

File: cerberus-webapp/src/main/java/ca/on/oicr/gsi/cerberus/ProvenanceServlet.java [hunk 4]
```diff
@@ -30,7 +30,7 @@
  *
  * @author ibancarz
  */
-@WebServlet(name = "ProvenanceServlet")
+@WebServlet(name = "ProvenanceServlet", urlPatterns = { "/provenance" })
```
Comment (apmasell) [inline cerberus-webapp/src/main/java/ca/on/oicr/gsi/cerberus/ProvenanceServlet.java:33]: Is this a reversion?

**Review by apmasell — COMMENTED** [review]
`log4j2.xml` and all the `WEB-INF` stuff need to be in `cerberus-webapp` not `cerberus-core`. The `log4j2.xml` might need to be duplicated in the `cerberus-cli`, but possibly configured differently.

File: cerberus-webapp/src/main/java/ca/on/oicr/gsi/cerberus/ProvenanceServlet.java (hunk 4 above)
Comment (callunity) [reply cerberus-webapp/src/main/java/ca/on/oicr/gsi/cerberus/ProvenanceServlet.java:33]: @iainrb what other servlets would be on that machine that would conflict with this one?

Comment (iainrb) [reply cerberus-webapp/src/main/java/ca/on/oicr/gsi/cerberus/ProvenanceServlet.java:33]: @apmasell Yes.
@callunity StatusServlet.java has:
`urlPatterns = { "/" }`
Which I believe is a conflict. At least it stopped the CLI from working with an installation on OpenStack.

I'm happy to hear other suggestions for resolving this without the `/provenance` suffix. :-)

Comment (apmasell) [reply cerberus-webapp/src/main/java/ca/on/oicr/gsi/cerberus/ProvenanceServlet.java:33]: Why is having the suffix a problem?

Comment (iainrb) [reply cerberus-webapp/src/main/java/ca/on/oicr/gsi/cerberus/ProvenanceServlet.java:33]: It was suggested to remove it in https://jira.oicr.on.ca/browse/GP-1862
But that was before the various reporting servlets were added, so the ticket may be obsolete now. Personally I'm happy to keep the `provenance` suffix if nobody objects.

Comment (apmasell) [reply cerberus-webapp/src/main/java/ca/on/oicr/gsi/cerberus/ProvenanceServlet.java:33]: Yeah. I think that the requirements of that ticket have changed.

Comment (iainrb) [reply cerberus-webapp/src/main/java/ca/on/oicr/gsi/cerberus/ProvenanceServlet.java:33]: Okay, I'll make a note on the ticket, and make the other changes you suggested above.

Comment (iainrb) [reply cerberus-webapp/src/main/java/ca/on/oicr/gsi/cerberus/ProvenanceServlet.java:33]: Latest commit hopefully includes all needed changes, let me know what you think.

**Review by apmasell — APPROVED** [review]

**Review by callunity — APPROVED** [review]

**Pull request merged**